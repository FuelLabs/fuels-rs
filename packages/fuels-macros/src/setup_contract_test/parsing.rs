use std::collections::HashSet;

use fuels_code_gen::ProgramType;
use itertools::{chain, Itertools};
use proc_macro2::Span;
use syn::{
    parse::{Parse, ParseStream},
    Error, LitStr, Result as ParseResult,
};

use crate::{
    abigen::MacroAbigenTarget,
    parse_utils::{Command, ErrorsExt, UniqueLitStrs, UniqueNameValues},
};

trait MacroCommand {
    fn expected_name() -> &'static str;
    fn validate_command_name(command: &Command) -> syn::Result<()> {
        let expected_name = Self::expected_name();
        if command.name == expected_name {
            Ok(())
        } else {
            Err(Error::new_spanned(
                command.name.clone(),
                format!("Expected command to have name: '{expected_name}'."),
            ))
        }
    }
}

pub(crate) struct InitializeWallet {
    pub(crate) span: Span,
    pub(crate) names: Vec<LitStr>,
}

impl MacroCommand for InitializeWallet {
    fn expected_name() -> &'static str {
        "Wallets"
    }
}

impl TryFrom<Command> for InitializeWallet {
    type Error = Error;

    fn try_from(command: Command) -> Result<Self, Self::Error> {
        Self::validate_command_name(&command)?;

        let wallets = UniqueLitStrs::new(command.contents)?;

        Ok(Self {
            span: command.name.span(),
            names: wallets.into_iter().collect(),
        })
    }
}

#[derive(Debug)]
pub(crate) struct AbigenCommand {
    pub(crate) span: Span,
    pub(crate) targets: Vec<MacroAbigenTarget>,
}

impl MacroCommand for AbigenCommand {
    fn expected_name() -> &'static str {
        "Abigen"
    }
}

impl TryFrom<Command> for AbigenCommand {
    type Error = Error;

    fn try_from(command: Command) -> Result<Self, Self::Error> {
        Self::validate_command_name(&command)?;

        let targets = command
            .contents
            .into_iter()
            .map(|meta| Command::new(meta).and_then(MacroAbigenTarget::new))
            .collect::<Result<Vec<_>, _>>()?;

        Ok(Self {
            span: command.name.span(),
            targets,
        })
    }
}

pub(crate) struct DeployContract {
    pub name: String,
    pub contract: LitStr,
    pub wallet: String,
}

impl MacroCommand for DeployContract {
    fn expected_name() -> &'static str {
        "Deploy"
    }
}

impl TryFrom<Command> for DeployContract {
    type Error = Error;

    fn try_from(command: Command) -> Result<Self, Self::Error> {
        Self::validate_command_name(&command)?;
        let name_values = UniqueNameValues::new(command.contents)?;
        name_values.validate_has_no_other_names(&["name", "contract", "wallet"])?;

        let name = name_values.get_as_lit_str("name")?.value();
        let contract = name_values.get_as_lit_str("contract")?.clone();
        let wallet = name_values.get_as_lit_str("wallet")?.value();

        Ok(Self {
            name,
            contract,
            wallet,
        })
    }
}

fn parse_test_contract_commands(
    input: ParseStream,
) -> syn::Result<(
    Vec<InitializeWallet>,
    Vec<AbigenCommand>,
    Vec<DeployContract>,
)> {
    let commands = Command::parse_multiple(input)?;

    let mut init_wallets: Vec<syn::Result<InitializeWallet>> = vec![];
    let mut gen_contracts: Vec<syn::Result<AbigenCommand>> = vec![];
    let mut deploy_contracts: Vec<syn::Result<DeployContract>> = vec![];

    let mut errors = vec![];

    for command in commands {
        let command_name = &command.name;
        if command_name == InitializeWallet::expected_name() {
            init_wallets.push(command.try_into());
        } else if command_name == AbigenCommand::expected_name() {
            gen_contracts.push(command.try_into());
        } else if command_name == DeployContract::expected_name() {
            deploy_contracts.push(command.try_into());
        } else {
            errors.push(Error::new_spanned(
                command.name,
                "Unsupported command. Expected: 'Wallets', 'Abigen' or 'Deploy'",
            ))
        }
    }

    let (init_wallets, wallet_errs): (Vec<_>, Vec<_>) = init_wallets.into_iter().partition_result();
    let (gen_contracts, gen_errs): (Vec<_>, Vec<_>) = gen_contracts.into_iter().partition_result();
    let (deploy_contracts, deploy_errs): (Vec<_>, Vec<_>) =
        deploy_contracts.into_iter().partition_result();

    chain!(errors, wallet_errs, gen_errs, deploy_errs).validate_no_errors()?;

    Ok((init_wallets, gen_contracts, deploy_contracts))
}

// Contains the result of parsing the input to the `setup_contract_test` macro.
// Contents represent the users wishes with regards to wallet initialization,
// bindings generation and contract deployment.
pub(crate) struct TestContractCommands {
    pub(crate) initialize_wallets: Option<InitializeWallet>,
    pub(crate) generate_bindings: AbigenCommand,
    pub(crate) deploy_contract: Vec<DeployContract>,
}

impl Parse for TestContractCommands {
    fn parse(input: ParseStream) -> ParseResult<Self> {
        let span = input.span();
        let (mut initialize_wallets, generate_contract, deploy_contract) =
            parse_test_contract_commands(input)?;

        let abigen_commands = Self::extract_the_abigen_command(span, generate_contract)?;

        Self::validate_all_contracts_are_known(&abigen_commands, deploy_contract.as_slice())?;

        Self::validate_zero_or_one_wallet_command_present(initialize_wallets.as_slice())?;

        Ok(Self {
            initialize_wallets: initialize_wallets.pop(),
            generate_bindings: abigen_commands,
            deploy_contract,
        })
    }
}

impl TestContractCommands {
    fn contracts_to_generate(commands: &AbigenCommand) -> HashSet<&LitStr> {
        commands
            .targets
            .iter()
            .filter_map(|target| match target.program_type {
                ProgramType::Contract => Some(&target.name),
                _ => None,
            })
            .collect()
    }

    fn contracts_to_deploy(commands: &[DeployContract]) -> HashSet<&LitStr> {
        commands.iter().map(|c| &c.contract).collect()
    }

    fn extract_the_abigen_command(
        parent_span: Span,
        mut commands: Vec<AbigenCommand>,
    ) -> Result<AbigenCommand, Error> {
        if commands.len() != 1 {
            let err = commands
                .iter()
                .map(|command| Error::new(command.span, "Only one `Abigen` command allowed"))
                .combine_errors()
                .unwrap_or_else(|| Error::new(parent_span, "Add an `Abigen(..)` command!"));

            Err(err)
        } else {
            Ok(commands.pop().unwrap())
        }
    }
    fn validate_all_contracts_are_known(
        generate_contracts: &AbigenCommand,
        deploy_contracts: &[DeployContract],
    ) -> syn::Result<()> {
        Self::contracts_to_deploy(deploy_contracts)
            .difference(&Self::contracts_to_generate(generate_contracts))
            .flat_map(|unknown_contract| {
                [
                    Error::new_spanned(unknown_contract, "Contract is unknown"),
                    Error::new(
                        generate_contracts.span,
                        format!(
                            "Consider adding: Contract(name=\"{}\", abi=...)",
                            unknown_contract.value()
                        ),
                    ),
                ]
            })
            .validate_no_errors()
    }

    fn validate_zero_or_one_wallet_command_present(
        commands: &[InitializeWallet],
    ) -> syn::Result<()> {
        if commands.len() > 1 {
            commands
                .iter()
                .map(|command| Error::new(command.span, "Only one `Wallets` command allowed"))
                .combine_errors()
                .map(Err)
                .expect("Known to have at least one error")
        } else {
            Ok(())
        }
    }
}
