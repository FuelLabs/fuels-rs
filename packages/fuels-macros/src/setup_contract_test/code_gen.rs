use std::{
    collections::HashMap,
    path::{Path, PathBuf},
};

use fuels_code_gen::{utils::ident, Abigen, AbigenTarget, ProgramType};
use proc_macro2::{Ident, TokenStream};
use quote::quote;
use rand::{prelude::StdRng, Rng, SeedableRng};
use syn::LitStr;

use crate::setup_contract_test::parsing::{
    DeployContract, GenerateContract, InitializeWallet, TestContractCommands,
};

pub(crate) fn generate_setup_contract_test_code(
    commands: TestContractCommands,
) -> syn::Result<TokenStream> {
    let TestContractCommands {
        initialize_wallets,
        generate_contract,
        deploy_contract,
    } = commands;

    let project_lookup = generate_project_lookup(&generate_contract)?;
    let abigen_code = abigen_code(&project_lookup);
    let wallet_code = wallet_initialization_code(initialize_wallets);
    let deploy_code = contract_deploying_code(&deploy_contract, &project_lookup);

    Ok(quote! {
       #abigen_code
       #wallet_code
       #deploy_code
    })
}

fn generate_project_lookup(commands: &[GenerateContract]) -> syn::Result<HashMap<String, Project>> {
    let pairs = commands
        .iter()
        .map(|command| -> syn::Result<_> {
            let project = Project::new(&command.abi)?;
            Ok((command.name.value(), project))
        })
        .collect::<Result<Vec<_>, _>>()?;

    Ok(pairs.into_iter().collect())
}

fn abigen_code(project_lookup: &HashMap<String, Project>) -> TokenStream {
    let targets = generate_abigen_targets(project_lookup);
    Abigen::generate(targets, false).expect("Failed to generate abigen")
}

fn generate_abigen_targets(project_lookup: &HashMap<String, Project>) -> Vec<AbigenTarget> {
    project_lookup
        .iter()
        .map(|(name, project)| AbigenTarget {
            name: name.clone(),
            abi: project.abi_path(),
            program_type: ProgramType::Contract,
        })
        .collect()
}

fn wallet_initialization_code(maybe_command: Option<InitializeWallet>) -> TokenStream {
    let command = if let Some(command) = maybe_command {
        command
    } else {
        return Default::default();
    };

    let wallet_names = extract_wallet_names(&command);

    if wallet_names.is_empty() {
        return Default::default();
    }

    let num_wallets = wallet_names.len();
    quote! {
        let [#(#wallet_names),*]: [_; #num_wallets] = launch_custom_provider_and_get_wallets(
            WalletsConfig::new(Some(#num_wallets as u64), None, None),
            None,
            None,
        )
        .await
        .try_into()
        .expect("Should have the exact number of wallets");
    }
}

fn extract_wallet_names(command: &InitializeWallet) -> Vec<Ident> {
    command
        .names
        .iter()
        .map(|name| ident(&name.value()))
        .collect()
}

fn contract_deploying_code(
    commands: &[DeployContract],
    project_lookup: &HashMap<String, Project>,
) -> TokenStream {
    commands
        .iter()
        .map(|command| {
            // Generate random salt for contract deployment
            let mut rng = StdRng::from_entropy();
            let salt: [u8; 32] = rng.gen();

            let contract_instance_name = ident(&command.name);
            let contract_struct_name = ident(&command.contract.value());
            let wallet_name = ident(&command.wallet);

            let project = project_lookup
                .get(&command.contract.value())
                .expect("Project should be in lookup");
            let bin_path = project.bin_path();
            let storage_path = project.storage_path();

            quote! {
                let #contract_instance_name = {
                    let storage_config = StorageConfiguration::load_from(#storage_path)
                                            .expect("Failed to load storage slots from path");
                    let load_config =
                        LoadConfiguration::default()
                            .set_storage_configuration(storage_config)
                            .set_salt([#(#salt),*]);

                    let loaded_contract = Contract::load_from(#bin_path, load_config).expect("Failed to load the contract");

                    let contract_id = loaded_contract.deploy(&#wallet_name, TxParameters::default()).await.expect("Failed to deploy the contract");
                    #contract_struct_name::new(contract_id, #wallet_name.clone())
                };
            }
        })
        .reduce(|mut all_code, code| {
            all_code.extend(code);
            all_code
        })
        .unwrap_or_default()
}

struct Project {
    path: PathBuf,
}

impl Project {
    fn new(dir: &LitStr) -> syn::Result<Self> {
        let path = Path::new(&dir.value()).canonicalize().map_err(|_| {
            syn::Error::new_spanned(
                dir.clone(),
                "Unable to canonicalize forc project path. Make sure the path is valid!",
            )
        })?;

        Ok(Self { path })
    }

    fn compile_file_path(&self, suffix: &str, description: &str) -> String {
        self.path
            .join(["out/debug/", self.project_name(), suffix].concat())
            .to_str()
            .unwrap_or_else(|| panic!("could not join path for {description}"))
            .to_string()
    }

    fn project_name(&self) -> &str {
        self.path
            .file_name()
            .expect("failed to get project name")
            .to_str()
            .expect("failed to convert project name to string")
    }

    fn abi_path(&self) -> String {
        self.compile_file_path("-abi.json", "the ABI file")
    }

    fn bin_path(&self) -> String {
        self.compile_file_path(".bin", "the binary file")
    }

    fn storage_path(&self) -> String {
        self.compile_file_path("-storage_slots.json", "the storage slots file")
    }
}
