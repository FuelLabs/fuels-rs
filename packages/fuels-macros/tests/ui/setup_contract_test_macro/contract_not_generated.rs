use fuels_macros::setup_contract_test;

setup_contract_test!(Deploy(
    name = "some_instance",
    contract = "SomeUnknownContract",
    wallet = "some_wallet"
));

fn main() {}
