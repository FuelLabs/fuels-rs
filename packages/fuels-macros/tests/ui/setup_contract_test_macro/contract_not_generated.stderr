error: Contract is unknown
 --> tests/ui/setup_contract_test_macro/contract_not_generated.rs:5:16
  |
5 |     contract = "SomeUnknownContract",
  |                ^^^^^^^^^^^^^^^^^^^^^

error: Consider adding: Abigen(name="SomeUnknownContract", abi=...)
 --> tests/ui/setup_contract_test_macro/contract_not_generated.rs:3:22
  |
3 | setup_contract_test!(Deploy(
  |                      ^^^^^^
